packages/react-query/src/shared/hooks/types.ts

```diff
@@ -138,10 +140,344 @@ export interface UseTRPCMutationOptions<
+  isStarting: false;
+  isStarted: true;
+  isConnecting: false;
+  isPending: true;
+  isReconnecting: false;
+  isError: false;
+  status: 'pending';
+}
+
+export const getPendingResult = <TInput, TError>(
+  previous: UseTRPCSubscriptionResult<TInput, TError>,
+): TRPCSubscriptionPendingResult<TInput, TError> => {
+  const time = Date.now();
+
+  if (previous.isStarting) {
+    return {
+      ...previous,
+      isStarting: false,
+      isStarted: true,
+      isConnecting: false,
+      isReconnecting: false,
+      isPending: true,
+      status: 'pending',
+      connectionStartedAt: time,
+      connectionAttemptCount: 0,
+      connectionError: null,
+      initialConnectionStartedAt: time,
+    };
+  }
+
+  return {
+    ...previous,
+    error: null,
+    isError: false,
+    isStarting: false,
+    isStarted: true,
+    isConnecting: false,
+    isReconnecting: false,
+    isPending: true,
+    status: 'pending',
+    connectionStartedAt: time,
+    connectionAttemptCount: 0,
+    connectionError: null,
+  };
+};
+
+export interface TRPCSubscriptionReconnectingResult<TInput, TError>
+  extends TRPCSubscriptionBaseResult<TInput, TError> {
+  error: null;
+  connectionError: TError;
+  isStarting: false;
+  isStarted: true;
+  isConnecting: true;
+  isPending: false;
+  isReconnecting: true;
+  isError: false;
+  status: 'connecting';
+}
+
+export const getReconnectingResult = <TInput, TError>(
+  previous:
+    | TRPCSubscriptionPendingResult<TInput, TError>
+    | TRPCSubscriptionReconnectingResult<TInput, TError>,
+  error: TError,
+): TRPCSubscriptionReconnectingResult<TInput, TError> => {
+  return {
+    ...previous,
+    isStarting: false,
+    isStarted: true,
+    isConnecting: true,
+    isReconnecting: true,
+    isPending: false,
+    status: 'connecting',
+    connectionError: error,
+    connectionAttemptCount: previous.connectionAttemptCount + 1,
+  };
+};
+
+export interface TRPCSubscriptionErrorResult<TInput, TError>
+  extends TRPCSubscriptionBaseResult<TInput, TError> {
+  error: TError;
+  isStarting: false;
+  isStarted: true; // Not sure about this one
+  isConnecting: false;
+  isPending: false;
+  isReconnecting: false;
+  isError: true;
+  status: 'error';
+}
+
+export const getErrorResult = <TInput, TError>(
+  previous: UseTRPCSubscriptionResult<TInput, TError>,
+  error: TError,
+): TRPCSubscriptionErrorResult<TInput, TError> => {
+  return {
+    ...previous,
+    isStarting: false,
+    isStarted: true,
+    isConnecting: false,
+    isReconnecting: false,
+    isPending: false,
+    isError: true,
+    status: 'error',
+    error,
+    errorUpdatedAt: Date.now(),
+  };
+};
+
+export type UseTRPCSubscriptionResult<TInput, TError> =
+  | TRPCSubscriptionIdleResult<TInput, TError>
+  | TRPCSubscriptionStartingResult<TInput, TError>
+  | TRPCSubscriptionPendingResult<TInput, TError>
+  | TRPCSubscriptionReconnectingResult<TInput, TError>
+  | TRPCSubscriptionErrorResult<TInput, TError>;
```

Comment:
Niiiiiiceeeee

cc @juliusmarminge what do you think of this approach to `.useSubscription()` results?